libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -1263,7 +1265,41 @@ bool ExpressionCompiler::visit(FunctionCall const& _functionCall)
 				// as require with custom errors is not supported in legacy codegen.
 				auto const* magicType = dynamic_cast<MagicType const*>(arguments[1]->annotation().type);
 				if (magicType && magicType->kind() == MagicType::Kind::Error)
-					solUnimplemented("Require with a custom error is only available using the via-ir pipeline.");
+				{
+					// Make sure that error constructor arguments are evaluated regardless of the require condition
+					auto const& errorConstructorCall = dynamic_cast<FunctionCall const&>(*arguments[1]);
+					errorConstructorCall.expression().accept(*this);
+					std::vector<Type const*> errorConstructorArgumentTypes{};
+					unsigned stackDepth = 0;
+					for (ASTPointer<Expression const> const& errorConstructorArgument: errorConstructorCall.sortedArguments())
+					{
+						errorConstructorArgument->accept(*this);
+						stackDepth += errorConstructorArgument->annotation().type->sizeOnStack();
+						errorConstructorArgumentTypes.push_back(errorConstructorArgument->annotation().type);
+					}
+					// stack: <condition> <arg0> <arg1> ... <argN>
+					// Move condition to the top of the stack
+					utils().moveIntoStack(arguments.at(0)->annotation().type->sizeOnStack(), stackDepth);
+					// stack: <arg0> <arg1> ... <argN> <condition>
+					m_context << Instruction::ISZERO << Instruction::ISZERO;
+					AssemblyItem successBranchTag = m_context.appendConditionalJump();
+
+					auto const* errorDefinition = dynamic_cast<ErrorDefinition const*>(ASTNode::referencedDeclaration(errorConstructorCall.expression()));
+					solAssert(errorDefinition && errorDefinition->functionType(true));
+
+					utils().revertWithError(
+						errorDefinition->functionType(true)->externalSignature(),
+						errorDefinition->functionType(true)->parameterTypes(),
+						errorConstructorArgumentTypes
+					);
```

Comment:
Is the internal function type really the one we want here? Why not external?

You should just get the type from the call, which will always be the right one:

```suggestion
					solAssert(dynamic_cast<ErrorDefinition const*>(ASTNode::referencedDeclaration(errorConstructor)));
					utils().revertWithError(
						errorConstructor.annotation().type.externalSignature(),
						errorConstructor.annotation().type.parameterTypes(),
						errorConstructorArgumentTypes
					);
```

Just add this variable above (there are other places in the handler that can use it too):
```c++
Expression const* errorConstructor = errorConstructorCall.expression();
```

Reply:
> Is the internal function type really the one we want here? Why not external?

Yes, we do the same thing for via-ir.

Also, `errorConstructorCall.expression.annotation().type` doesn't have `externalSignature()`, nor `parameterTypes()`.

Reply:
This does work, but looks like `errorConstructor.annotation().type` just needs a cast to `FunctionType`. I suggested this because this is actually how we do it in the other place in `ExpressionCompiler` where we call `revertWithError()` (`function` is already of `FunctionType` there though):

https://github.com/ethereum/solidity/blob/7711a8ecd0cc28e25b01aeabbd8c3459c7b5b234/libsolidity/codegen/ExpressionCompiler.cpp#L1001-L1005

In the end it's not a big issue, they both hopefully give you the same thing, but accessing this via the `FunctionCall` passed into your handler just makes more sense to me. Reaching for the type via definition just makes me want to assert something because I feel like I could be getting a different thing. And on top of that I have to think if I need an internal or external part.